### PR TITLE
Refactor data conversion errors and diagnostics, tests, headers and test data directory

### PR DESCRIPTION
**Pull request by mkgrgis**
In this PR:
- PostgreSQL `integer`-family _out of range_ checks.  SQLite `int` affinity is 1↔1 with PostgreSQL `bigint` and also have equal ranges. If we have `int` or `smallint` in PostgreSQL some input values which is out of range for this data types are possible.
- `sqlite3_column_@(stmt, colid)` functions inside of [`sqlite_convert_to_pg`](https://github.com/pgspider/sqlite_fdw/blob/a387b2f49ddba00fb59fad76de56e8a8df4b7d1a/sqlite_query.c#L51) replaced to `sqlite3_value_@(val)` functions for less parameters and fastest calculations, see [`sqlite3_value *sqlite3_column_value(sqlite3_stmt *pStmt, int i`)](https://github.com/sqlite/sqlite/blob/4c6554223eacb131092506cc423606ec6977c21e/src/vdbeapi.c#L1403) and [`static Mem *columnMem(sqlite3_stmt *pStmt, int i)`](https://github.com/sqlite/sqlite/blob/4c6554223eacb131092506cc423606ec6977c21e/src/vdbeapi.c#L1308) which will calculated once instead of many `sqlite3_column_@(stmt, colid)` calls.
- Detach `out_of_range` test from `int4`, add `int2` TCs https://github.com/pgspider/sqlite_fdw/pull/91/commits/641c539615b36ff203199b0c908524cf069ba358.
- In some cases in `bigint` PostgreSQL column can come a value with `real` affinity which is out of range. Add PostgreSQL input core function and TCs for this case.
- Deep refactoring of conversion data error diagnostics to newly implemented `ErrorContextCallback` function with universal messages about any data conversion problems between PostgreSQL data column and SQLite data value.
- Detach `bitstring` test from `type` test https://github.com/pgspider/sqlite_fdw/pull/91/commits/9bc53a470d0a51d40e34e4dfe6d25ad3e45fd3d3.
- Add `dtof` to `float4` PostgreSQL data column branch https://github.com/pgspider/sqlite_fdw/pull/91/commits/f0593dead92aad86ff27e7bea307984a4f3eeb97 (without testing, will tested in other PR about full `real` affinity support).
- Move SQLite test data to separate `/tmp/sqlite_fdw_test` directory for less chaos in `/tmp` https://github.com/pgspider/sqlite_fdw/pull/91/commits/cbba0432a2824c21912dc45f7b760ab4e350d54e . This allow reduce testing time.
- Refactor headers: remove unnecessary headers, ABC other, limit to some versions some of rare used headers. This allow reduce compile time. https://github.com/pgspider/sqlite_fdw/pull/91/commits/e23ba835b19ac894f20bdb56d289f4615db0795d and https://github.com/pgspider/sqlite_fdw/pull/91/commits/932b3e0a0bc69e04cc7cd8b00089d4d51915d25a
- Encapsulate `ForeignTable` and `ForeignServer` into `...State` structures, because this variables are often calculated near this structures in functions like constructor.  https://github.com/pgspider/sqlite_fdw/pull/91/commits/f1a1f89f733dc9f256c1c482018538c9964abdbe

### CODE REVIEW

**Comment by mkgrgis** [comment]
@t-kataym, this simple PR is ready to review.

**Comment by mkgrgis** [comment]
@t-kataym , could you please write approximate week of review of this PR? Have pgspider team any plans of a new `sqlite_fdw` release?

File: sql/12.16/extra/int4.sql [hunk 1]
```diff
@@ -369,6 +369,28 @@ INSERT INTO INT4_TMP VALUES (2147483647::int4, 2147483646::int4);
...
+ALTER FOREIGN TABLE INT4_TMP ALTER COLUMN f1 TYPE int8;
+--Testcase 137:
+INSERT INTO INT4_TMP VALUES (x'7FFFFFFF'::int8 + 1, 0);
+--Testcase 138:
+ALTER FOREIGN TABLE INT4_TMP ALTER COLUMN f1 TYPE int4;
+--Testcase 139:
+SELECT * FROM INT4_TMP; -- overflow
+
+--Testcase 140:
+DELETE FROM INT4_TMP;
```
Comment (nxhai98) [inline sql/12.16/extra/int4.sql:384]: Could you add test code for int2 overflow?

File: sqlite_query.c [hunk 2]
```diff
@@ -145,8 +152,12 @@ sqlite_convert_to_pg(Form_pg_attribute att, sqlite3_stmt * stmt, int stmt_colid,
 				{
 					case SQLITE_INTEGER: /* <-- proper and recommended SQLite affinity of value for pgtyp */
 					{
-						int			value = sqlite3_column_int(stmt, stmt_colid);
-						return (struct NullableDatum) {Int32GetDatum(value), false};
+						sqlite_int64 i64v = sqlite3_column_int64(stmt, stmt_colid);
+						if (i64v > PG_INT32_MAX || i64v < PG_INT32_MIN)
```
Comment (nxhai98) [inline sqlite_query.c:156]: What about int64 overflow? It happens when we insert too big number (PG_INT64_MAX + 1) to SQLite column, SQLite will store it as a double value: 

> If the TEXT value is a well-formed integer literal that is too large to fit in a 64-bit signed integer, it is converted to REAL. https://www.sqlite.org/datatype3.html

I think we can check simply by
```C
case INT8OID:
{
		switch (sqlite_value_affinity)
		{
				....
				case SQLITE_DOUBLE:
				{
				        double		value = sqlite3_column_double(stmt, stmt_colid);
				        Datum     d = DirectFunctionCall1(dtoi8, Float8GetDatum((float8) value));
				        return (struct NullableDatum) {d, false};
				}
....
```

**Review by nxhai98 — COMMENTED** [review]
@mkgrgis,

Thank for your contribution.
I have some comments for your PR, could you please check it?

File: sqlite_query.c [hunk 3]
```diff
@@ -118,8 +121,12 @@ sqlite_convert_to_pg(Form_pg_attribute att, sqlite3_stmt * stmt, int stmt_colid,
 				{
 					case SQLITE_INTEGER: /* <-- proper and recommended SQLite affinity of value for pgtyp */
 						{
-							int			value = sqlite3_column_int(stmt, stmt_colid);
-							return (struct NullableDatum) {Int16GetDatum(value), false};
+							sqlite_int64 i64v = sqlite3_column_int64(stmt, stmt_colid);
+							if (i64v > PG_INT16_MAX || i64v < PG_INT16_MIN)
```
[the remark this answers is not in the document]
Comment (mkgrgis) [reply sqlite_query.c:125]: Thanks for hint, @nxhai98! Your variant really is more flexible. I'll replace the code fragment and change tests.

[the remark this answers is not in the document]
Comment (mkgrgis) [reply sqlite_query.c:125]: @nxhai98 , PostgreSQL core functions doesn't gives a value caused error. In my `ereport` there was `errhint("Value %s in column \"%.*s\"", text_value, (int)sizeof(pgColND.data), pgColND.data)));` Why do you think core functions is better for SQLite data case where usually there is many bad-prepared or not normal datasets?

[the remark this answers is not in the document]
Comment (nxhai98) [reply sqlite_query.c:125]: @mkgrgis,

> Why do you think core functions is better for SQLite data case where usually there are many bad-prepared or not normal datasets?

The data conversion process (not for int conversion) should be done by PostgreSQL core function, we can avoid dependency of binary presentation of PostgeSQL in sqlite_fdw. Some data have complex range checking/validation e.g. float8 to float4 conversion cannot implement simply by compare value:
```C
	result = (float4) num;
	if (unlikely(isinf(result)) && !isinf(num))
		float_overflow_error();
	if (unlikely(result == 0.0f) && num != 0.0)
		float_underflow_error();

	PG_RETURN_FLOAT4(result);
```

> PostgreSQL core functions doesn't gives a value caused error.

For error detail, we can add more error context/detail by using `ErrorContextCallback`, referred from postgres_fdw implementation: https://github.com/postgres/postgres/blob/278972f5933679e06e7d71d2060c1867f5aa4ebe/contrib/postgres_fdw/postgres_fdw.c#L7671

File: sqlite_query.c [hunk 4]
```diff
@@ -118,8 +121,12 @@ sqlite_convert_to_pg(Form_pg_attribute att, sqlite3_stmt * stmt, int stmt_colid,
...
+							{
+								sqlite_numeric_overflow (att, stmt, attnum);
+							}
+							return (struct NullableDatum) {Int16GetDatum((int)i64v), false};
 						}
 					case SQLITE_FLOAT:
```
[the remark this answers is not in the document]
Comment (mkgrgis) [reply sqlite_query.c:131]: Unfortunately this implementation planned in other PR after this and also `dtof` usage cause to analyse many TCs in `float4` test. I am not ready for `float4` works now.

File: sql/12.16/extra/int4.sql [hunk 5]
```diff
@@ -369,6 +369,28 @@ INSERT INTO INT4_TMP VALUES (2147483647::int4, 2147483646::int4);
 --Testcase 134:
 SELECT lcm(f1, f2) FROM INT4_TMP; -- overflow
 
+--Testcase 135:
+DELETE FROM INT4_TMP;
```
[the remark this answers is not in the document]
Comment (mkgrgis) [reply sql/12.16/extra/int4.sql]: Moved in https://github.com/pgspider/sqlite_fdw/pull/91/commits/641c539615b36ff203199b0c908524cf069ba358. Will also change overflow check to core functions in next commit.

File: sql/12.16/extra/int4.sql (hunk 1 above)
Comment (mkgrgis) [reply sql/12.16/extra/int4.sql]: Added in https://github.com/pgspider/sqlite_fdw/commit/641c539615b36ff203199b0c908524cf069ba358. Will rewritten to PostgreSQL core functions in next commit.

File: sqlite_query.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (mkgrgis) [reply sqlite_query.c:125]: Many thanks for `ErrorContextCallback`, @nxhai98 ! Initial implementation allow improve error messages in `encodings` tests.

[the remark this answers is not in the document]
Comment (mkgrgis) [reply sqlite_query.c]: Core functions added and `ErrorContextCallback` function implemented in https://github.com/pgspider/sqlite_fdw/pull/91/commits/06ee232ebf1785ea5daed0391f5aa12d890b6dd6 . Please, verify @nxhai98.

File: sqlite_query.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (mkgrgis) [reply sqlite_query.c:138]: `dtof` added in https://github.com/pgspider/sqlite_fdw/pull/91/commits/f0593dead92aad86ff27e7bea307984a4f3eeb97, no testing problems detected in current commit chain. Please verify, @nxhai98.

File: sqlite_query.c (hunk 2 above)
Comment (mkgrgis) [reply sqlite_query.c]: Added in https://github.com/pgspider/sqlite_fdw/pull/91/commits/06ee232ebf1785ea5daed0391f5aa12d890b6dd6 without testing. `float` data for `bigint`/`int8` PostgreSQL column is not normal.

**Comment by mkgrgis** [comment]
> I have some comments for your PR, could you please check it?

All your comments have checked, please start 2nd round of review, @nxhai98 !

File: expected/12.16/extra/bool.out [hunk 6]
```diff
@@ -249,8 +254,13 @@ INSERT INTO "type_BOOLEAN"(i, b) VALUES (27, 3.14159265358979);
 ALTER FOREIGN TABLE "type_BOOLEAN" ALTER COLUMN "b" TYPE bool;
 --Testcase 49: ERR - invalid float for bool column
 SELECT * FROM "type_BOOLEAN+";
-ERROR:  SQLite data affinity "real" disallowed for PostgreSQL data type "boolean"
-HINT:  In column "b" expected SQLite affinity "integer", incorrect value = '3.14159265358979'
+ERROR:  SQLite value is not compatible with PostgreSQL column data type
+CONTEXT:  foreign table "type_BOOLEAN+"
```
Comment (nxhai98) [inline expected/12.16/extra/bool.out:258]: This message seems too detail, how do you think about:
```
ERROR:  SQLite value is not compatible with PostgreSQL column data type
CONTEXT:  column "b" of foreign table "type_BOOLEAN+"
HINT:  In column "b" expected SQLite affinity "integer", incorrect value = '3.14159265358979'
```

File: sqlite_fdw.c [hunk 7]
```diff
@@ -5690,3 +5718,251 @@ sqliteIsForeignRelUpdatable(Relation rel)
...
+	int32	 	pgtypmod = att->atttypmod;
+	NameData	pgColND = att->attname;
+	const char *pg_dataTypeName = NULL;
+	const char *sqlite_affinity = NULL;
+	const char *pg_good_affinity = NULL;
+	const int	max_logged_byte_length = NAMEDATALEN * 2;
+	int 		value_byte_size_blob_or_utf8 = sqlite3_value_bytes (errpos->val);
+	int			value_aff = sqlite3_value_type(errpos->val);
+	char	   *value_text = NULL;
+	char	   *err_cont_mess0 = malloc(4 * NAMEDATALEN + max_logged_byte_length * 2 + 1024); /* The longest context message */
```
Comment (nxhai98) [inline sqlite_fdw.c:5750]: please using `palloc` instead of `malloc`

File: sqlite_query.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (nxhai98) [reply sqlite_query.c:134]: Thank you. Please help me resolve this thread.

File: expected/12.16/extra/bitstring.out [hunk 8]
```diff
@@ -0,0 +1,810 @@
+--SET log_min_messages  TO DEBUG1;
```
Comment (nxhai98) [inline expected/12.16/extra/bitstring.out:1]: Is this come from PR #95?

File: expected/16.0/extra/bool.out [hunk 9]
```diff
@@ -4,7 +4,7 @@
 CREATE EXTENSION sqlite_fdw;
 --Testcase 001:
 CREATE SERVER sqlite_svr FOREIGN DATA WRAPPER sqlite_fdw
-OPTIONS (database '/tmp/sqlitefdw_test.db');
+OPTIONS (database '/tmp/sqlite_fdw_test/common.db');
```
Comment (nxhai98) [inline expected/16.0/extra/bool.out:7]: Could you explain about this change?

**Review by nxhai98 — COMMENTED** [review]
I have some more comment, could you please check it?

File: sqlite_query.c (hunk 2 above)
Comment (nxhai98) [reply sqlite_query.c]: Thank you.

Could you add test for this code. Leaving a new source code untested is not a good practice.

File: sql/12.16/extra/int4.sql (hunk 5 above)
[the remark this answers is not in the document]
Comment (nxhai98) [reply sql/12.16/extra/int4.sql]: Thank you. please help me resolve this thread.

File: sqlite_query.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (nxhai98) [reply sqlite_query.c]: Thank you. please help me resolve this thread.

File: expected/12.16/extra/bitstring.out (hunk 8 above)
Comment (mkgrgis) [reply expected/12.16/extra/bitstring.out:1]: Yes, if here was big test changes around of conversion error `CONTEXT` there is reason make all formal test normalisation here.

File: expected/16.0/extra/bool.out (hunk 9 above)
Comment (mkgrgis) [reply expected/16.0/extra/bool.out:7]: It's for 
- simpler test script, just `testdir='/tmp/sqlite_fdw_test'; rm -rf "$testdir";`
- less chaos in `/tmp`, because I have many programs with files in `/tmp`
- up to 1.5% faster tests, maybe because `/tmp/sqlite_fdw_test` inodes can be better cached than `/tmp` with many different files from different processes. Also `/tmp/sqlite_fdw_test` content have no changes not from `sqlite_fdw` during testing.

File: sqlite_fdw.c (hunk 7 above)
Comment (mkgrgis) [reply sqlite_fdw.c]: Fixed in https://github.com/pgspider/sqlite_fdw/pull/91/commits/91fa86abe3c795766fb1981db75512062d1b557b , please verify @nxhai98

File: expected/12.16/extra/bool.out (hunk 6 above)
Comment (mkgrgis) [reply expected/12.16/extra/bool.out:258]: Yes, @nxhai98, not all of this is `CONTEXT`. `errhint` also can work in this function. I think any PostgreSQL data context elements we can leave in `CONTEXT`, but SQLite value details can move to `HINT`, are you agree?

File: sqlite_query.c (hunk 2 above)
Comment (mkgrgis) [reply sqlite_query.c]: Hm... This code is not a good practice at all  ;-) because we can lost not integer part of the value. I'll add overflow TC in `out_of_range` tests, but it's better to think how we can exclude this `INT8OID` + `SQLITE_DOUBLE` branch needed only in SQLite `mod()` context.

Comment (mkgrgis) [reply sqlite_query.c]: Here is no overflow, @nxhai98. There is very bad behaviour. What should I test here?
```
ALTER FOREIGN TABLE INT8_TBL ALTER COLUMN q1 TYPE double precision;
INSERT INTO INT8_TBL VALUES (-9223372036854775809, 0); // https://www.postgresql.org/docs/current/datatype-numeric.html#DATATYPE-NUMERIC
+ALTER FOREIGN TABLE INT8_TBL ALTER COLUMN q1 TYPE int8;
+SELECT * FROM INT8_TBL;
+          q1          |        q2         
+----------------------+-------------------
+ -9223372036854775808 |                 0
```
We have -922337203685477580**9** in `INSERT`, but  -922337203685477580**8** in `SELECT` and no out or range error.

File: expected/12.16/extra/bool.out (hunk 6 above)
Comment (mkgrgis) [reply expected/12.16/extra/bool.out:259]: Refactored in https://github.com/pgspider/sqlite_fdw/pull/91/commits/6a70cc999e37b76b2eb93ea030f5b11ab587c10e based on `CONTEXT` is more detailed and about stable PostgreSQL environment, but `HINT` is a message with unstable format about SQLite value, which can printable or not, quotable or not, original or hexadeciamal, with or without useful data length.

File: sqlite_query.c (hunk 2 above)
Comment (nxhai98) [reply sqlite_query.c]: It's because of floating point precision, `-9223372036854775808` and `-9223372036854775809` have same binary presentation in float type => the error will not happen. If we use `PG_INT64_MAX + 1` and higher or much lower than PG_INT64_MIN like `10 * PG_INT64_MIN`, the out of range error will be happen.

File: sqlite_fdw.c [hunk 10]
```diff
@@ -5690,3 +5718,279 @@ sqliteIsForeignRelUpdatable(Relation rel)
...
+		{
+			if (sqlite_value_as_hex_code)
+				err_hint_mess += sprintf(
+						err_hint_mess,
+						"SQLite hex value %s\n",
+						value_text );
+			else if (value_aff != SQLITE_INTEGER && value_aff != SQLITE_FLOAT)
+				err_hint_mess += sprintf(
+						err_hint_mess,
+						"SQLite value '%s'\n",
```
Comment (nxhai98) [inline sqlite_fdw.c:5920]: Could you remove `\n`? I think it does not need to break line in hint/context message.

File: sqlite_fdw.c [hunk 11]
```diff
@@ -5690,3 +5718,279 @@ sqliteIsForeignRelUpdatable(Relation rel)
...
+		/*
+		 * Error CONTEXT block
+		 */
+		char	   *err_cont_mess0 = palloc(4 * NAMEDATALEN + 64); /* The longest context message */
+		char 	   *err_cont_mess;
+
+		err_cont_mess = err_cont_mess0;
+		err_cont_mess = err_cont_mess + sprintf(
+			err_cont_mess,
+			"foreign table \"%s\"\nforeign column \"%.*s\" have data type \"%s\" (usual affinity \"%s\")\n",
```
Comment (nxhai98) [inline sqlite_fdw.c:5962]: How do you think about message: `"column \"%.*s\" of foreign table \"%s\" have data type \"%s\" (usual affinity \"%s\")"`

Comment (mkgrgis) [reply sqlite_fdw.c:5962]: I am agree. Really `\n` is not necessary in `CONTEXT` part of error message.

File: sqlite_fdw.c (hunk 10 above)
Comment (mkgrgis) [reply sqlite_fdw.c:5920]: SQLite value can be long (up to `NAMEDATALEN * 4`  in _hex_ case), hence additional part of hint (affinity and sometimes data length) will not very useful for reading. How do you think, @nxhai98 ?

Comment (nxhai98) [reply sqlite_fdw.c]: I think it no problem, `NAMEDATALEN * 4` in default config just have about 256 characters (3~4 terminal line) and it may an unusual case.   
At most cases in our test, this line to too short.

File: sqlite_query.c (hunk 2 above)
Comment (mkgrgis) [reply sqlite_query.c]: Thanks for clarification, @nxhai98 , changed to -9223372036854775810 in https://github.com/pgspider/sqlite_fdw/pull/91/commits/c79a5281848396153a582386e9fd79ac7678d855 and no problems.

Comment (mkgrgis) [reply sqlite_query.c]: Also added `10 * PG_INT64_MIN`  TCs in https://github.com/pgspider/sqlite_fdw/pull/91/commits/ddd6a141af7ecf2db9a38ae19c97f270bfe4689c and leaved TCs with problematic behaviour between `PG_INT64_MIN` and `10 * PG_INT64_MIN`.

File: sqlite_fdw.c (hunk 10 above)
Comment (mkgrgis) [reply sqlite_fdw.c]: Thanks for idea, @nxhai98 !  A good readable version without `\n`: `SQLite value with "" affinity` than `(%d bytes)` if applicable, than `in hex` if printable in hex and the original long value at the end. Changed to this message in https://github.com/pgspider/sqlite_fdw/pull/91/commits/311b818fdde67cd0bc5b45ac600d7359cfcc6d6f, please verify.

File: sqlite_fdw.c (hunk 11 above)
Comment (mkgrgis) [reply sqlite_fdw.c]: Also additional words about query from reference `postgres_fdw` implementation added without `\n` in https://github.com/pgspider/sqlite_fdw/pull/91/commits/311b818fdde67cd0bc5b45ac600d7359cfcc6d6f. Please verify, @nxhai98.

File: sqlite_query.c (hunk 2 above)
Comment (nxhai98) [reply sqlite_query.c]: Thank you. Please help me resolve this thread.

File: expected/12.16/extra/bool.out (hunk 6 above)
Comment (nxhai98) [reply expected/12.16/extra/bool.out]: Thank you. Please help me resolve this thread.

File: sqlite_fdw.c (hunk 7 above)
Comment (nxhai98) [reply sqlite_fdw.c]: Thank you. Please help me resolve this thread.

File: expected/12.16/extra/bitstring.out (hunk 8 above)
Comment (nxhai98) [reply expected/12.16/extra/bitstring.out:1]: Is this test just moved from type.sql without any modify?
If so, there is no problem. Please help me resolve this thread.

File: expected/16.0/extra/bool.out (hunk 9 above)
Comment (nxhai98) [reply expected/16.0/extra/bool.out:7]: Thank you. Please help me resolve this thread.

File: sqlite_fdw.c (hunk 10 above)
Comment (nxhai98) [reply sqlite_fdw.c]: Thank you. Please help me resolve this thread.

File: sqlite_fdw.c (hunk 11 above)
Comment (nxhai98) [reply sqlite_fdw.c]: Thank you. Please help me resolve this thread.

File: expected/12.16/extra/bitstring.out (hunk 8 above)
Comment (mkgrgis) [reply expected/12.16/extra/bitstring.out:1]: Yes, @nxhai98 , this is just moved test without any TC modifying. Resolved.

**Comment by mkgrgis** [comment]
All comments are checked and fixed, @nxhai98. Thanks for review! What will be later? 3rd round of your review or review by @t-kataym ?

**Comment by mkgrgis** [comment]
Ping, @nxhai98 ! Are there any results of 2 rounds of your review? What will be later?

**Comment by mkgrgis** [comment]
@t-kataym , could you please write approximate week of review and status of this PR? Will there 3rd review round?

**Comment by mkgrgis** [comment]
Hello @t-kataym and @nxhai98 ! Any chances to have feedback or 3rd review round for this PR? What is the status of this PR? Completely PR description after checking all your comments was updated https://github.com/pgspider/sqlite_fdw/pull/91#issue-1994632771

**Comment by t-kataym** [comment]
@mkgrgis , Thank you for  your contribution.  
@nxhai98 , Thank you for your confirmation.  
I will merge the PR.
